File: src/main/scala/data/MessageLogRepository.scala
package data

import domain.AlertUpdate
import doobie.ConnectionIO
import doobie.implicits._

trait MessageLogRepository {
  def insertMessage(message: AlertUpdate): ConnectionIO[Int]
}

class MessageLogRepositoryImpl extends MessageLogRepository {
  def insertMessage(message: AlertUpdate): ConnectionIO[Int] =
    sql"""insert into messages_log (message, hash) values (${message.content}, ${message.hash})
          on conflict (hash) do nothing"""
      .update
      .run
}

object MessageLogRepository {
  def apply(): MessageLogRepository = new MessageLogRepositoryImpl()
}

File: src/OrbWeaver.Data/Repositories/MessageLogRepository.cs
using Microsoft.EntityFrameworkCore;
using OrbWeaver.Domain;
using OrbWeaver.Handler.Abstractions;

namespace OrbWeaver.Data.Repositories;

public class MessageLogRepository(IDbContextFactory<OrbWeaverDbContext> contextFactory) : IMessageLogRepository
{
    public async Task<int> Log(Message message, CancellationToken cancellationToken = default)
    {
        await using var context = await contextFactory.CreateDbContextAsync(cancellationToken);
        return await context.Database
            .ExecuteSqlInterpolatedAsync
            ($"""
              insert into messages_log(hash, payload) 
              values ({message.Hash}, {message.Payload})
              on conflict (hash) do nothing;
            """, cancellationToken: cancellationToken);
    }
}